core/src/modules/ops/destination.ts

```diff
@@ -912,13 +913,45 @@ export namespace DestinationOps {
     success: boolean;
     error: Error;
   }> {
-    const _exports: Export[] = []; // only ones we are sending
-    for (const _export of givenExports) {
-      // TODO: maybe should recalcuate hasChanges and from current mostRecent
-      if (!_export.hasChanges) {
-        await _export.complete(); // do not include exports with hasChanges=false
+    const _exports: Export[] = [];
+    const locks: Lock[] = [];
+
+    for (const givenExport of givenExports) {
+      const lock = await getOrSetLock({
+        key: `${givenExport.recordId}:${givenExport.destinationId}`,
+      });
+      locks.push(lock);
+
+      if (lock.isLocked) {
+        // is _export the newest export for this pair?  if not, cancel it.
+        const mostRecentExport = await Export.findOne({
+          where: {
+            recordId: givenExport.recordId,
+            destinationId: givenExport.destinationId,
+            state: "pending",
+          },
+          order: ["createdAt", "ASC"],
+        });
+        //if the export we're looking at is older than the one that locked this pair, cancel it
+        if (mostRecentExport.createdAt > givenExport.createdAt) {
```

Comment:
So if I read this right, we couldn't get the lock (because an export is being processsed in another thread or earlier in this batch) _and_ now we check against other pending exports. That right?

If the `mostRecentExport` is newer, I understand cancelling this one and `continue`.
But if `givenExport` is newer, we will still want to try it later. I think it should still `continue` though because letting it fall through like it is now means we still end up with them in parallel. right?

Reply:
I did what you suggested below, checking for a `moreRecentExport` and if one exists, cancelling the one we're looking at.  I think that should address the case you're talking about here, yes?

Reply:
But if `givenExport` is the most recent and `lock.isLocked`, doesn't that mean that we are currently processing and older export that got the lock? In that case, we should wait to process `givenExport` until later. As it is now, it still gets processed in parallel with the one that got the lock.

It's all very tricky, so let me know if I'm wrong.

Reply:
Nope that sounds right.  I had a `continue` in there earlier for that case...  It should have stayed.